From 50db4d958b87459602397879966b1ed2ea628f55 Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@solus-project.com>
Date: Fri, 14 Jul 2017 15:10:58 +0100
Subject: [PATCH 26/37] Make daddy proud and do a visual refresh

Long story short, we now prefer coloured icons to symbolic ones, we
have a sidebar which lists all possible navigation, and it looks kinda
pretty.

Now people actually know how many steps are involved. *Gasp*.

Signed-off-by: Ikey Doherty <ikey@solus-project.com>
---
 os_installer2/data/styling.css       |   6 ++
 os_installer2/mainwindow.py          | 115 ++++++++++++++++++++++-------------
 os_installer2/pages/basepage.py      |  15 +++++
 os_installer2/pages/complete.py      |   3 +
 os_installer2/pages/disk_location.py |   5 +-
 os_installer2/pages/geoip.py         |   5 +-
 os_installer2/pages/keyboard.py      |   6 +-
 os_installer2/pages/language.py      |   5 +-
 os_installer2/pages/location.py      |   5 +-
 os_installer2/pages/partitioning.py  |   7 ++-
 os_installer2/pages/progress.py      |   5 +-
 os_installer2/pages/summary.py       |   3 +
 os_installer2/pages/system.py        |   7 ++-
 os_installer2/pages/timezone.py      |   5 +-
 os_installer2/pages/users.py         |   5 +-
 15 files changed, 144 insertions(+), 53 deletions(-)

diff --git a/os_installer2/data/styling.css b/os_installer2/data/styling.css
index 6305de4..582b0eb 100644
--- a/os_installer2/data/styling.css
+++ b/os_installer2/data/styling.css
@@ -7,3 +7,9 @@ scrolledwindow {
     background-image: none;
     background-color: transparent;
 }
+
+.header-box {
+    background-color: #222D32;
+    box-shadow: 0 2px 3px 1px alpha(black, 0.47);
+    color: white;
+}
diff --git a/os_installer2/mainwindow.py b/os_installer2/mainwindow.py
index 9d56f57..b07c7da 100644
--- a/os_installer2/mainwindow.py
+++ b/os_installer2/mainwindow.py
@@ -30,6 +30,19 @@ import sys
 import threading
 import traceback
 
+class FancyLabel(Gtk.Label):
+
+    page_id = None
+
+    def __init__(self, page):
+        Gtk.Label.__init__(self)
+        self.set_label(page.get_sidebar_title())
+        self.page_id = page.get_name()
+        self.set_halign(Gtk.Align.START)
+        self.set_property("margin", 6)
+        self.set_property("margin-start", 24)
+        self.set_property("margin-end", 24)
+        self.get_style_context().add_class("dim-label")
 
 class InstallInfo:
     """ For tracking purposes between pages """
@@ -89,6 +102,8 @@ class MainWindow(Gtk.ApplicationWindow):
     prev_button = None
     next_button = None
 
+    box_labels = None
+
     pages = list()
     page_index = 0
 
@@ -116,59 +131,77 @@ class MainWindow(Gtk.ApplicationWindow):
         Gtk.ApplicationWindow.__init__(self, application=app)
         self.application = app
 
-        self.headerbar = Gtk.HeaderBar.new()
-        self.headerbar.set_show_close_button(True)
-        self.set_titlebar(self.headerbar)
+        Gtk.Settings.get_default().set_property("gtk-application-prefer-dark-theme", False)
 
-        self.image_step = Gtk.Image.new_from_icon_name("system-software-install", Gtk.IconSize.LARGE_TOOLBAR)
-        self.image_step.set_pixel_size(32)
-        self.image_step.set_property("margin", 4)
+        self.image_step = Gtk.Image.new_from_icon_name("system-software-install", Gtk.IconSize.DIALOG)
+        self.image_step.set_property("margin", 8)
         self.label_step = Gtk.Label.new("")
+        self.label_step.set_property("margin", 8)
+
+        self.set_title("Install Solus")
+
+        self.headerbox = Gtk.Box.new(Gtk.Orientation.VERTICAL, 0)
+        self.headerbox.get_style_context().add_class("header-box")
+        self.headerbox.pack_start(self.image_step, False, False, 0)
+        # self.headerbox.pack_start(self.label_step, False, False, 0)
+        self.box_labels = Gtk.Box.new(Gtk.Orientation.VERTICAL, 0)
+        self.box_labels.set_valign(Gtk.Align.START)
+        self.box_labels.set_property("margin-bottom", 40)
+        self.box_labels.set_property("margin-top", 20)
+        self.headerbox.pack_start(self.box_labels, True, True, 0)
+
+        # Vanity! TODO: Select correct icon ..
+        img_vanity = Gtk.Image.new_from_icon_name("budgie-desktop_badge-symbolic", Gtk.IconSize.LARGE_TOOLBAR)
+        img_vanity.set_property("margin", 8)
+        img_vanity.set_property("margin-top", 0)
+        lab_vanity = Gtk.Label.new("Solus Budgie")
+        lab_vanity.set_property("margin-start", 4)
+        lab_vanity.set_property("margin-end", 8)
+        lab_vanity.set_property("margin-bottom", 8)
+        self.headerbox.pack_end(lab_vanity, False, False, 0)
+        self.headerbox.pack_end(img_vanity, False, False, 0)
 
-        box = Gtk.Box.new(Gtk.Orientation.HORIZONTAL, 0)
-        box.pack_start(self.label_step, False, False, 0)
-        self.headerbar.set_custom_title(box)
-        self.headerbar.pack_start(self.image_step)
-
-        self.set_title("")
         self.set_icon_name("system-software-install")
         self.connect("delete-event", self.quit_handler)
 
         self.set_position(Gtk.WindowPosition.CENTER)
         self.set_default_size(768, 500)
 
-        # Main view
-        self.stack = Gtk.Stack()
-        ltr = Gtk.StackTransitionType.SLIDE_LEFT_RIGHT
-        self.stack.set_transition_type(ltr)
-        self.add(self.stack)
-
         # Main "install" page
         self.installer_page = Gtk.VBox(0)
         self.installer_stack = Gtk.Stack()
         self.installer_page.pack_start(self.installer_stack, True, True, 0)
 
+        self.installer_wrap = Gtk.Box.new(Gtk.Orientation.HORIZONTAL, 0)
+        self.installer_wrap.pack_start(self.headerbox, False, False, 0)
+        sep = Gtk.Separator.new(Gtk.Orientation.VERTICAL)
+        self.installer_wrap.pack_start(sep, False, False, 0)
+        self.installer_wrap.pack_start(self.installer_page, True, True, 0)
+
+        ltr = Gtk.StackTransitionType.SLIDE_LEFT_RIGHT
         self.installer_stack.set_transition_type(ltr)
-        self.stack.add_named(self.installer_page, "install")
+        self.add(self.installer_wrap)
+
 
         # nav buttons
-        bbox = Gtk.ButtonBox.new(Gtk.Orientation.HORIZONTAL)
-        bbox.set_layout(Gtk.ButtonBoxStyle.SPREAD)
+        bbox = Gtk.Box.new(Gtk.Orientation.HORIZONTAL, 4)
         bbox.set_halign(Gtk.Align.END)
-        self.installer_page.pack_end(bbox, False, 0, 0)
         self.prev_button = Gtk.Button.new_with_label("Previous")
-        self.prev_button.get_style_context().add_class("flat")
         self.next_button = Gtk.Button.new_with_label("Next")
-        self.next_button.get_style_context().add_class("flat")
-        self.next_button.get_style_context().add_class("suggested-action")
-        bbox.add(self.prev_button)
-        bbox.add(self.next_button)
-        bbox.set_margin_bottom(20)
-        bbox.set_margin_end(30)
-        bbox.set_margin_top(20)
+        bbox.pack_start(self.prev_button, False, False, 0)
+        bbox.pack_start(self.next_button, False, False, 0)
+        bbox.set_margin_top(10)
+        bbox.set_margin_bottom(10)
+        bbox.set_margin_end(10)
         self.prev_button.set_property("margin-start", 4)
         self.next_button.set_property("margin-start", 4)
 
+        # sep before nav
+        sep = Gtk.Separator.new(Gtk.Orientation.HORIZONTAL)
+        sep.set_margin_top(20)
+        self.installer_page.pack_end(bbox, False, 0, 0)
+        self.installer_page.pack_end(sep, False, 0, 0)
+
         self.info = InstallInfo()
         self.info.owner = self
 
@@ -226,17 +259,12 @@ class MainWindow(Gtk.ApplicationWindow):
         self.set_can_next(True)
         Gdk.threads_leave()
 
-    def phase_install(self):
-        self.stack.set_visible_child_name("install")
-
-    def phase_live(self):
-        """ Consider switching to another view showing how to restart the
-            installer ? """
-        self.application.quit()
 
     def add_installer_page(self, page):
         """ Work a page into the set """
         self.installer_stack.add_named(page, page.get_name())
+        lab = FancyLabel(page)
+        self.box_labels.pack_start(lab, False, False, 0)
         self.pages.append(page)
 
     def next_page(self):
@@ -290,11 +318,16 @@ class MainWindow(Gtk.ApplicationWindow):
         else:
             self.set_can_previous(True)
         page.prepare(self.info)
-        mk = u"<span font-size='x-large'>{}</span>".format(page.get_title())
-        self.label_step.set_markup(mk)
+
+        for label in self.box_labels.get_children():
+            if label.page_id == page.get_name():
+                label.get_style_context().remove_class("dim-label")
+            else:
+                label.get_style_context().add_class("dim-label")
+
         self.image_step.set_from_icon_name(page.get_icon_name(),
-                                           Gtk.IconSize.LARGE_TOOLBAR)
-        self.image_step.set_pixel_size(32)
+                                           Gtk.IconSize.DIALOG)
+        # self.image_step.set_pixel_size(32)
         self.installer_stack.set_visible_child_name(page.get_name())
 
     def set_can_previous(self, can_prev):
diff --git a/os_installer2/pages/basepage.py b/os_installer2/pages/basepage.py
index eb7f984..d97d113 100644
--- a/os_installer2/pages/basepage.py
+++ b/os_installer2/pages/basepage.py
@@ -20,9 +20,24 @@ class BasePage(Gtk.VBox):
     def __init__(self):
         Gtk.VBox.__init__(self)
 
+        mk = u"<span font-size='x-large'>{}</span>".format(self.get_title())
+        lab = Gtk.Label.new(mk)
+        lab.set_property("margin-top", 10)
+        lab.set_property("margin-start", 20)
+        lab.set_property("margin-bottom", 10)
+        lab.set_use_markup(True)
+        lab.set_halign(Gtk.Align.START)
+        self.pack_start(lab, False, False, 0)
+
+        sep = Gtk.Separator.new(Gtk.Orientation.HORIZONTAL)
+        self.pack_start(sep, False, False, 0)
+
     def get_title(self):
         return None
 
+    def get_sidebar_title(self):
+        return "Not implemented.."
+
     def get_name(self):
         return None
 
diff --git a/os_installer2/pages/complete.py b/os_installer2/pages/complete.py
index 5593c10..e65b12d 100644
--- a/os_installer2/pages/complete.py
+++ b/os_installer2/pages/complete.py
@@ -62,6 +62,9 @@ class InstallationCompletePage(BasePage):
     def get_name(self):
         return "complete"
 
+    def get_sidebar_title(self):
+        return "Complete"
+
     def get_icon_name(self):
         return "start-here-solus"
 
diff --git a/os_installer2/pages/disk_location.py b/os_installer2/pages/disk_location.py
index 93a433a..148363d 100644
--- a/os_installer2/pages/disk_location.py
+++ b/os_installer2/pages/disk_location.py
@@ -226,11 +226,14 @@ class InstallerDiskLocationPage(BasePage):
     def get_title(self):
         return "Where should we install?"
 
+    def get_sidebar_title(self):
+        return "Disks"
+
     def get_name(self):
         return "disk-location"
 
     def get_icon_name(self):
-        return "drive-harddisk-system-symbolic"
+        return "disk-utility"
 
     def load_disks(self):
         """ Load the disks within a thread """
diff --git a/os_installer2/pages/geoip.py b/os_installer2/pages/geoip.py
index e303802..35f521b 100644
--- a/os_installer2/pages/geoip.py
+++ b/os_installer2/pages/geoip.py
@@ -54,8 +54,11 @@ class InstallerGeoipPage(BasePage):
     def get_name(self):
         return "geoip-lookup"
 
+    def get_sidebar_title(self):
+        return u"↠" + " Find location"
+
     def get_icon_name(self):
-        return "find-location-symbolic"
+        return "mark-location-symbolic"
 
     def prepare(self, info):
         self.info = info
diff --git a/os_installer2/pages/keyboard.py b/os_installer2/pages/keyboard.py
index ed897e0..d93f954 100644
--- a/os_installer2/pages/keyboard.py
+++ b/os_installer2/pages/keyboard.py
@@ -237,11 +237,15 @@ class InstallerKeyboardPage(BasePage):
     def get_title(self):
         return "Choose a keyboard layout"
 
+    def get_sidebar_title(self):
+        return "Keyboard"
+
     def get_name(self):
         return "keyboard"
 
     def get_icon_name(self):
-        return "input-keyboard-symbolic"
+        # Just looks nicer than input-keyboard
+        return "preferences-desktop-keyboard-shortcuts"
 
     def prepare(self, info):
         self.info = info
diff --git a/os_installer2/pages/language.py b/os_installer2/pages/language.py
index 580f795..eb90949 100644
--- a/os_installer2/pages/language.py
+++ b/os_installer2/pages/language.py
@@ -122,8 +122,11 @@ class InstallerLanguagePage(BasePage):
     def get_title(self):
         return "Choose a language"
 
+    def get_sidebar_title(self):
+        return "Language"
+
     def get_name(self):
         return "language"
 
     def get_icon_name(self):
-        return "preferences-desktop-locale-symbolic"
+        return "preferences-desktop-locale"
diff --git a/os_installer2/pages/location.py b/os_installer2/pages/location.py
index 34d2548..23ced4f 100644
--- a/os_installer2/pages/location.py
+++ b/os_installer2/pages/location.py
@@ -51,11 +51,14 @@ are in the world and proceed automatically.")
     def get_title(self):
         return "Where are you?"
 
+    def get_sidebar_title(self):
+        return "Location"
+
     def get_name(self):
         return "location"
 
     def get_icon_name(self):
-        return "find-location-symbolic"
+        return "maps"
 
     def prepare(self, info):
         self.info = info
diff --git a/os_installer2/pages/partitioning.py b/os_installer2/pages/partitioning.py
index 2bc9b9d..cc77f34 100644
--- a/os_installer2/pages/partitioning.py
+++ b/os_installer2/pages/partitioning.py
@@ -727,13 +727,16 @@ class InstallerPartitioningPage(BasePage):
         self.stack.set_visible_child_name("automatic")
 
     def get_title(self):
-        return "Configure disks"
+        return "Advanced disk configuration"
+
+    def get_sidebar_title(self):
+        return "Disk Configuration"
 
     def get_name(self):
         return "partitioning"
 
     def get_icon_name(self):
-        return "drive-multidisk-symbolic"
+        return "disk-usage-analyzer"
 
     def prepare(self, info):
         self.info = info
diff --git a/os_installer2/pages/progress.py b/os_installer2/pages/progress.py
index fe436db..0848090 100644
--- a/os_installer2/pages/progress.py
+++ b/os_installer2/pages/progress.py
@@ -135,7 +135,10 @@ class InstallerProgressPage(BasePage):
         return "install"
 
     def get_icon_name(self):
-        return "install-symbolic"
+        return "system-software-install"
+
+    def get_sidebar_title(self):
+        return "Install Solus"
 
     def begin_install(self):
         """ Begin the real work of doing the installation """
diff --git a/os_installer2/pages/summary.py b/os_installer2/pages/summary.py
index b1e13c8..a7b40b8 100644
--- a/os_installer2/pages/summary.py
+++ b/os_installer2/pages/summary.py
@@ -86,6 +86,9 @@ class InstallerSummaryPage(BasePage):
         items.pack_start(self.system_details, False, False, 2)
 
     def get_title(self):
+        return "Review options before installation"
+
+    def get_sidebar_title(self):
         return "Summary"
 
     def get_name(self):
diff --git a/os_installer2/pages/system.py b/os_installer2/pages/system.py
index fff88b4..f44d638 100644
--- a/os_installer2/pages/system.py
+++ b/os_installer2/pages/system.py
@@ -139,13 +139,16 @@ class InstallerSystemPage(BasePage):
         self.check_forward()
 
     def get_title(self):
-        return "System Settings"
+        return "Configure the bootloader &amp; hostname"
 
     def get_name(self):
         return "system-settings"
 
+    def get_sidebar_title(self):
+        return "System Settings"
+
     def get_icon_name(self):
-        return "preferences-other-symbolic"
+        return "preferences-system"
 
     def prepare(self, info):
         self.info = info
diff --git a/os_installer2/pages/timezone.py b/os_installer2/pages/timezone.py
index 8a0046c..6b0fee7 100644
--- a/os_installer2/pages/timezone.py
+++ b/os_installer2/pages/timezone.py
@@ -68,11 +68,14 @@ class InstallerTimezonePage(BasePage):
     def get_title(self):
         return "Choose your timezone"
 
+    def get_sidebar_title(self):
+        return "Timezone"
+
     def get_name(self):
         return "timezone"
 
     def get_icon_name(self):
-        return "preferences-system-time-symbolic"
+        return "preferences-system-time"
 
     def change_timezone(self, completion, model, selection):
         item = model[selection]
diff --git a/os_installer2/pages/users.py b/os_installer2/pages/users.py
index 7efeed8..c13fe6d 100644
--- a/os_installer2/pages/users.py
+++ b/os_installer2/pages/users.py
@@ -356,8 +356,11 @@ class InstallerUsersPage(BasePage):
     def get_title(self):
         return "Who will use this device?"
 
+    def get_sidebar_title(self):
+        return "Users"
+
     def get_name(self):
         return "users"
 
     def get_icon_name(self):
-        return "system-users-symbolic"
+        return "system-users"
-- 
2.13.3

